### PR TITLE
MAINT: utilise `array_api_compat` v1.8

### PR DESCRIPTION
**Pull request by lucascolley**
#### What does this implement/fix?
- Updates `array_api_compat` submodule to v1.8 (released today)
- Remove our custom helpers `xp_minimum` and `xp_clip`

### CODE REVIEW

File: scipy/stats/_stats_py.py
Comment (lucascolley) [inline scipy/stats/_stats_py.py:1522]: erroneous `xp=xp`

**Comment by rgommers** [comment]
Thanks Lucas! Diff looks good so far; some issue with `test_mstats_basic.py::TestVariability::test_zmap` that CI complains about.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4627]:
```suggestion
```

**Comment by lucascolley** [comment]
can you tell what's causing the `zmap` failures @mdhaber ?

**Comment by mdhaber** [comment]
Yeah `xp_asarray` is not passing masked arrays through.

**Comment by lucascolley** [comment]
thanks, I'll look into it

File: scipy/stats/_morestats.py
Comment (mdhaber) [inline scipy/stats/_morestats.py:4096]:
```suggestion
    R = xp.clip(hypotenuse, max=1.)
```

File: scipy/stats/_morestats.py
Comment (mdhaber) [inline scipy/stats/_morestats.py:4200]:
```suggestion
    R = xp.clip(hypotenuse, max=1.)  # [1] (2.2.4)
```

**Review by mdhaber — COMMENTED** [review]
Not really important here, but I happened to notice that this would be more idiomatic (and compact) for this use case.

File: scipy/_lib/_array_api.py
Comment (mdhaber) [inline scipy/_lib/_array_api.py:169]: Any guesses what this was for?

Comment (lucascolley) [reply scipy/_lib/_array_api.py:169]: historical cruft I believe - looks like it initially always used NumPy regardless of array type, then later this was guarded for only NumPy arrays (my guess)

File: tools/check_test_name.py
Comment (mdhaber) [inline tools/check_test_name.py:75]: What does this do?

**Comment by lucascolley** [comment]
For me locally, this lets the `masked_array`'s pass through, although the tests now have tolerance errors.

File: tools/check_installation.py
Comment (mdhaber) [inline tools/check_installation.py:47]: and this?

**Review by mdhaber — APPROVED** [review]
If CI is happy, LGTM after https://github.com/scipy/scipy/pull/21264/files#r1693153699 is committed (`[skip ci]` would be OK).

File: tools/check_test_name.py
Comment (lucascolley) [reply tools/check_test_name.py:75]: Checks that test function names start with `test` so that they are detected properly by `pytest`. I looked at what this was doing in `array_api_compat` and it looks intentional (not a typo). We could do with a better way of ignoring submodules in this script though. cc @j-bowhay @MarcoGorelli

**Comment by mdhaber** [comment]
Well, now `xp_mean` and `xp_var` (maybe others) still use `xp.asarray` on the input in case it is, for example, a list. But this makes it not a masked array. We could use `_asarray` there with `subok=True`, I suppose.

**Comment by lucascolley** [comment]
> I suppose this brings up the question of whether xp.asarray was supposed to be preserving subclasses before? Should it still?

I don't think it has to. Subclasses are out of scope for the standard [^1], so I think the behaviour is undefined.

[^1]: https://data-apis.org/array-api/2023.12/purpose_and_scope.html#out-of-scope

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:74]: Why change formatting and not import as `xp_array` anymore?

Comment (lucascolley) [reply scipy/stats/_stats_py.py:74]: Where was the `xp_array`? I've just added the import of `_asarray`, and imported `size` from `_array_api` rather than `array_api_compat`.

The formatting change is just to improve readability. IMO this is a useful thing to do when the lines are already being touched (while we wait for auto-formatting to materialise), but I can see the argument for keeping the diff as minimal as possible. In this case, it only takes a few seconds to check.

Comment (mdhaber) [reply scipy/stats/_stats_py.py:74]: My concern is that argument can be made for any change that formats a la Black, which is very polarizing. My opinions about it are mixed - if we adopted it as a project, I'd understand and be OK with it. But we haven't yet, so I'd prefer not to set a precedent that unnecessary Black format changes are ok. You're a maintainer, too, so you can do it if you want, but I might still say something about it.

I thought `_asarray` was already imported as `xp_asarray`, but I guess not. I do try to do that when I import things. You know, for "readability" : )

Comment (lucascolley) [reply scipy/stats/_stats_py.py:74]: I'm happy to review or submit a PR to change `_asarray` to `xp_asarray`! I think I chose `_asarray` to be as close to `asarray` as possible, but actually I do agree that `xp_asarray` is more readable.

> I might still say something about it.

Me too :) until we have a project-wide decision I think the sensible thing is to accept either way. I think there is a substantial difference between somebody sending PRs specifically for opinionated style cleanups which we haven't agreed upon (indeed, we would not want to set a precedent for that), and a contributor making the imports of the files they are working on easier for them to read (when the lines are already diffed).

**Comment by lucascolley** [comment]
I'm rather confused by the CI failures as I can't reproduce them locally.

**Comment by rgommers** [comment]
> > I suppose this brings up the question of whether xp.asarray was supposed to be preserving subclasses before? Should it still?
> 
> I don't think it has to. Subclasses are out of scope for the standard [1](#user-content-fn-1-ec16b0b3274e594a61932c1c3a1a3bee), so I think the behaviour is undefined (no mention in [the `xp.asarray` docs](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.asarray.html)).

In general: the standard has no opinion on this indeed. Some array libraries do not allow subclassing at all. Of the ones that do, only NumPy's behavior has legacy subclass issues I think, that's why `np.asarray` disallows subclasses and `np.asanyarray` exists. 

I'd expect that our `_asarray` will be matching `np.asarray` by default for numpy arrays, not `np.asanyarray`.

File: scipy/stats/_stats_py.py
Comment (rgommers) [reply scipy/stats/_stats_py.py:74]: Agreed with the `black` concern - nicer to not change styles.

> I'm happy to review or submit a PR to change `_asarray` to `xp_asarray`! I think I chose `_asarray` to be as close to `asarray` as possible, but actually I do agree that `xp_asarray` is more readable.

I will note that we discussed this name twice already now, first when naming it `as_xparray`, then with the rename to `_asarray`. See for example https://github.com/scipy/scipy/issues/18866#issuecomment-1634998051.

In general, our `xp_func` names are matching behavior with the `xp.func` equivalent, and we're putting them into `_lib` when things aren't yet available upstream etc. This function on the other hand has a lot of extra keywords and semantics, and does not match `xp.asarray`. 

So that's two reasons for leaving it as is - that'd be better imho.

**Review by tylerjereddy — COMMENTED** [review]
> I'm rather confused by the CI failures as I can't reproduce them locally.

I can repro locally with: `python dev.py test -t scipy/special/tests/test_support_alternative_backends.py -b all`

I bisected `array-api-compat` on the namespace return change from `scipy._lib.array_api_compat.numpy` to plain `numpy` and found:

```
a7da56023f0e4e21d5b53e9eba1806688d09a43b is the first bad commit
commit a7da56023f0e4e21d5b53e9eba1806688d09a43b
Author: Aaron Meurer <asmeurer@gmail.com>
Date:   Wed Mar 27 16:55:16 2024 -0600

    Make use_compat a public flag to array_namespace()
    
    If None, it does the default behavior. If True or False it forces the use or
    non-use of the compat wrapper.
    
    This commit also changes it so that NumPy 2.0 does not return the wrapper
    library at all by default (when use_compat=None), since it is completely array
    API compatible on its own.

 array_api_compat/common/_helpers.py | 44 +++++++++++++++++++++++++++++--------
 tests/_helpers.py                   |  7 ++++--
 tests/test_array_namespace.py       | 23 +++++++++++--------
 3 files changed, 54 insertions(+), 20 deletions(-)
```

That makes sense I guess, given NumPy 2.x compat improvements. I haven't thought too hard about what we want to do here, but the patch below allows the test command above to pass locally. It looks a bit ugly, but anyway I suspect this information is somewhat helpful in making a decision on what might be wrong or what patch might be appropriate:

```diff
--- a/scipy/_lib/_array_api.py
+++ b/scipy/_lib/_array_api.py
@@ -126,7 +126,10 @@ def array_namespace(*arrays: Array) -> ModuleType:
 
     _arrays = compliance_scipy(_arrays)
 
-    return array_api_compat.array_namespace(*_arrays)
+    try:
+        return array_api_compat.array_namespace(*_arrays, use_compat=True)
+    except ValueError: # i.e., array_api_strict
+        return array_api_compat.array_namespace(*_arrays, use_compat=None)
```

**Comment by lucascolley** [comment]
Thanks Tyler, super helpful!

> I'm rather confused by the CI failures as I can't reproduce them locally.

Ah, I should probably have NumPy 2.0 in my dev environment by now...

Here is a clearer picture of where the problem lies:

```shell
$ SCIPY_ARRAY_API=1 python dev.py ipython

In [1]: import numpy as xp

In [2]: from scipy._lib._array_api import array_namespace

In [3]: array_namespace(0.0)
Out[3]: <module 'numpy' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/__init__.py'>

In [4]: zero_Celsius = 273.15

In [5]: array_namespace(zero_Celsius)
Out[5]: <module 'numpy' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/__init__.py'>

In [6]: _val = xp.asarray(32.)

In [7]: array_namespace(_val)
Out[7]: <module 'numpy' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/__init__.py'>

In [8]: tempo = (_val - 32) * 5 / 9 + zero_Celsius

In [9]: array_namespace(tempo)
Out[9]: <module 'scipy._lib.array_api_compat.numpy' from '/Users/lucascolley/programming/scipy/build-install/lib/python3.10/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

```

**Comment by lucascolley** [comment]
Hmm, unfortunately NumPy 2.0.1 has enough incompatibilities that this is causing quite a few errors. I think I'm going to pass `use_compat=True` for now, but FYI @asmeurer this default might cause a bit of confusion if NumPy releases can't stay ahead of the standard / array-api-compat.

**Comment by asmeurer** [comment]
Right, numpy 2.0 needs to be wrapped for 2023 support. The dream of having a completely unwrapped NumPy didn't last very long unfortunately.

File: scipy/_lib/_array_api.py
Comment (lucascolley) [inline scipy/_lib/_array_api.py:132]: https://github.com/data-apis/array-api-compat/issues/168

**Review by lucascolley — COMMENTED** [review]
CI is green and this is ready to go IMO!

File: scipy/_lib/_array_api.py
Comment (rgommers) [reply scipy/_lib/_array_api.py:132]: If we want to merge this PR before resolving this, it'd be good to add a code comment about `try:` to refer to the upstream issue.

Comment (lucascolley) [reply scipy/_lib/_array_api.py:133]: done!

**Comment by lucascolley** [comment]
I don't think this needs to be squashed.

**Comment by lucascolley** [comment]
merge conflicts resolved

File: scipy/_lib/_array_api.py
Comment (lucascolley) [reply scipy/_lib/_array_api.py:133]: upstream PR is in, so this can be removed before merge

Comment (rgommers) [reply scipy/_lib/_array_api.py:133]: Want to make the change now? I'm also fine with doing it as a follow-up if you prefer.

File: scipy/stats/_stats_py.py
Comment (rgommers) [reply scipy/stats/_stats_py.py:74]: Re the reformatting: I think the point made about not overdoing `black`-style reformatting is valid and good to keep in mind. But now that the change is made here and the code certainly doesn't look worse (arguably more readable indeed), keeping this change seems fine to me.

File: tools/check_test_name.py
Comment (rgommers) [reply tools/check_test_name.py:75]: That's a long-standing and minor issue, so let's ignore it here (a PR is always welcome of course).

**Review by rgommers — APPROVED** [review]
Changes all LGTM now, this seems mergeable as is - fine with me to do that now, or update for the try-except removal and bumping the submodule again (any preference @lucascolley?).

**Comment by lucascolley** [comment]
ready to go if CI is green!

**Comment by lucascolley** [comment]
failures in the Windows `npPre` job look unrelated

**Comment by rgommers** [comment]
Merged, thanks Lucas & reviewers!

**Pull request merged**
